aidial_adapter_openai/gpt4_multi_modal/transformation.py

```diff
@@ -0,0 +1,129 @@
+from typing import List, Optional, cast
+
+from pydantic import BaseModel, root_validator
+
+from aidial_adapter_openai.gpt4_multi_modal.download import (
+    ImageFail,
+    download_image,
+)
+from aidial_adapter_openai.gpt4_multi_modal.image_tokenizer import (
+    tokenize_image,
+)
+from aidial_adapter_openai.gpt4_multi_modal.messages import (
+    create_image_message,
+    create_text_message,
+)
+from aidial_adapter_openai.utils.image_data_url import ImageDataURL
+from aidial_adapter_openai.utils.log_config import logger
+from aidial_adapter_openai.utils.storage import FileStorage
+from aidial_adapter_openai.utils.tokens import Tokenizer
+
+
+class MessageTransformResult(BaseModel):
+    text_tokens: int = 0
+    image_tokens: int = 0
+    errors: Optional[List[ImageFail]] = None
+    message: dict
+
+    @property
+    def total_tokens(self) -> int:
+        return self.text_tokens + self.image_tokens
+
+
+class TransformMessagesResult(BaseModel):
```

Comment:
There are two ways to name thing:
1. Reflect in its name how it was produced - to answer the question _What is it the result of_?
2. Reflect what is it semantically - _What does it represent? How could it be used downstream?_

First hightlights past of an object, second - its future.

The second is more commonly used, because we rathely care about the past of objects.

So maybe `GPT4OMessages` would be a better name.

Reply:
Came up with `MultiModalMessage` naming, since we can use this truncation not only for GPT40, but for GPT4Vision also